include/nbl/system/ICancellableAsyncQueueDispatcher.h

```diff
@@ -97,108 +106,124 @@ class ICancellableAsyncQueueDispatcherBase
-    using base_t = IAsyncQueueDispatcher<CRTP, RequestType, BufferSize, InternalStateType>;
-    friend base_t;
+        using this_async_queue_t = ICancellableAsyncQueueDispatcher<CRTP, RequestType, BufferSize, InternalStateType>;
+        using base_t = IAsyncQueueDispatcher<CRTP, RequestType, BufferSize, InternalStateType>;
+        friend base_t;
+
+        template <typename T>
+        class future_storage_t
+        {
+            public:
+                alignas(T) uint8_t storage[sizeof(T)];
+
+                T* getStorage() { return reinterpret_cast<T*>(storage); }
+        };
 
-    template <typename T>
-    class future_storage_t
-    {
     public:
-        alignas(T) uint8_t storage[sizeof(T)];
+        using request_base_t = impl::ICancellableAsyncQueueDispatcherBase::request_base_t;
 
-        T* getStorage() { return reinterpret_cast<T*>(storage); }
-    };
+        static_assert(std::is_base_of_v<request_base_t, RequestType>, "Request type must derive from request_base_t!");
 
-public:
-    using request_base_t = impl::ICancellableAsyncQueueDispatcherBase::request_base_t;
+        template <typename T>
+        class future_t : private future_storage_t<T>, public impl::ICancellableAsyncQueueDispatcherBase::future_base_t
+        {
+                friend this_async_queue_t;
 
-    static_assert(std::is_base_of_v<request_base_t, RequestType>, "Request type must derive from request_base_t!");
+                // construct the retval element 
+                template <typename... Args>
+                void notify(Args&&... args)
+                {
+                    new (future_storage_t<T>::getStorage()) T(std::forward<Args>(args)...);
+                    valid_flag.store(true);
+                }
 
-    template <typename T>
-    class future_t : private future_storage_t<T>, public impl::ICancellableAsyncQueueDispatcherBase::future_base_t
-    {
-        friend this_async_queue_t;
+                //! See ICancellableAsyncQueueDispatcher::associate_request_with_future() docs
+                void associate_request(RequestType* req)
+                {
+                    request = req;
+                }
 
-        // construct the retval element 
-        template <typename... Args>
-        void notify(Args&&... args)
-        {
-            new (future_storage_t<T>::getStorage()) T(std::forward<Args>(args)...);
-            valid_flag.store(true);
-        }
+            public:
+                using value_type = T;
 
-        //! See ICancellableAsyncQueueDispatcher::associate_request_with_future() docs
-        void associate_request(RequestType* req)
-        {
-            request = req;
-        }
+                void cancel()
+                {
+                    bool valid = valid_flag.exchange(false);
+                    if (valid)
+                        future_storage_t<T>::getStorage()->~T();
+                    impl::ICancellableAsyncQueueDispatcherBase::future_base_t::cancel();
```

Comment:
i believe this function should call `set_cancel` on request?

Reply:
`ICancellableAsyncQueueDispatcherBase::future_base_t::cancel` does exactly that

but I looked at the code and it seems like it should cancel before it tries to delete the storage instead of what its doing now and was doing before (not my bug)

Reply:
hm, yeah, probbaly would be better to first cancel and then delete the storage